### PR TITLE
Fixed launching when the containing path contains a ^ or &.

### PR DESCRIPTION
**Pull request by 333fred**
Internal bug #262666. This introduces a basic string parsing state machine, which does not quote any characters that are inside a quoted string. Tagging @BillHiebert @dotnet/project-system for review.

Tagging @MattGertz for ask mode approval.

### CODE REVIEW

**Comment by davkean** [comment]
Can we keep functionality changes separate from formatting in future? :)

**Comment by 333fred** [comment]
Sorry. I have VS set up to autotrim trailing whitespace.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Debug/ConsoleDebugTargetsProvider.cs [hunk 1]
```diff
@@ -321,24 +316,103 @@ public string GetFullPathOfExeFromEnvironmentPath(string exeToSearchFor)
...
+        /// <param name="unescaped">The string to escape.</param>
+        /// <param name="toEscape">The characters to escape in the string.</param>
+        /// <returns>The escaped string.</returns>
+        internal static string EscapeString(string unescaped, char[] toEscape)
+        {
+            if (string.IsNullOrWhiteSpace(unescaped)) return unescaped;
+
+            var currentState = StringState.NoSlash;
+            var finalBuilder = new StringBuilder();
+            foreach (var currentLetter in unescaped)
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Debug/ConsoleDebugTargetsProvider.cs]: Nit, this is a "char" not a letter, there's a difference.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Debug/ConsoleDebugTargetsProvider.cs [hunk 2]
```diff
@@ -321,24 +316,103 @@ public string GetFullPathOfExeFromEnvironmentPath(string exeToSearchFor)
...
+                            currentState = StringState.InString;
+                        }
+                        else if (toEscape.Contains(currentLetter))
+                        {
+                            finalBuilder.Append('^');
+                        }
+
+                        finalBuilder.Append(currentLetter);
+                        break;
+                    case StringState.OneSlash:
```
Comment (333fred) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Debug/ConsoleDebugTargetsProvider.cs]: The only thing I'm a little wondering about is whether or not this is actually a valid state. Does `\"` have special meaning in the shell? A quick test of `echo \"` prints `\"`, so I think this is actually incorrect.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Debug/ConsoleDebugTargetsProvider.cs [hunk 3]
```diff
@@ -321,24 +316,103 @@ public string GetFullPathOfExeFromEnvironmentPath(string exeToSearchFor)
...
+                        else if (currentLetter == '"')
+                        {
+                            currentState = StringState.NoSlash;
+                        }
+
+                        finalBuilder.Append(currentLetter);
+                        break;
+                    case StringState.InStringOneSlash:
+                        // If we have one slash, then we blindly append to the string, no escaping, and move back to
+                        // InString.
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Debug/ConsoleDebugTargetsProvider.cs]: I find it helps to give examples, for example, took a little to understand that this is handling when you escape a quote within a quoted string.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Debug/ConsoleDebugTargetsProvider.cs (hunk 2 above)
Comment (333fred) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Debug/ConsoleDebugTargetsProvider.cs]: Nevermind. A quick C# test app proved that this is necessary. `.\countArgs \" asdf \"` will print 3 as the number of arguments. This state is necessary.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Debug/ConsoleDebugTargetsProvider.cs [hunk 4]
```diff
@@ -321,24 +316,103 @@ public string GetFullPathOfExeFromEnvironmentPath(string exeToSearchFor)
...
+                        finalBuilder.Append(currentLetter);
+                        currentState = StringState.InString;
+                        break;
+                    default:
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Debug/ConsoleDebugTargetsProvider.cs]: Nit: I do the following to get rid of the dead code in the default case:

``` C#
case StringState.InStringOneSlash:
case default:
Assumes.True(state == StringState.InStringOneSlash);

```

Comment (333fred) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Debug/ConsoleDebugTargetsProvider.cs]: But that replaces a section of dead code with an extra function call that will get hit every time that case gets hit. Doesn't seem like a worthwhile tradeoff to me.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Debug/ConsoleDebugTargetsProvider.cs [hunk 5]
```diff
@@ -321,24 +316,103 @@ public string GetFullPathOfExeFromEnvironmentPath(string exeToSearchFor)
...
+                        // The valid transitions are to OneSlash and InString. Any '"' encountered here is a raw quote.
+                        if (currentLetter == '\\')
+                        {
+                            currentState = StringState.OneSlash;
+                        }
+                        else if (currentLetter == '"')
+                        {
+                            currentState = StringState.InString;
+                        }
+                        else if (toEscape.Contains(currentLetter))
```
Comment (333fred) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Debug/ConsoleDebugTargetsProvider.cs]: Changed toEscape to a statically allocated list.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Debug/ConsoleDebugTargetsProvider.cs [hunk 6]
```diff
@@ -10,24 +10,29 @@
...
+    /// CPS debugger but hooks into the launch profiles extensibility point. The order of this provider is
     /// near the bottom to ensure other providers get chance to handle it first
     /// </summary>
     [Export(typeof(IDebugProfileLaunchTargetsProvider))]
     [AppliesTo(ProjectCapability.LaunchProfiles)]
     [ExportOrder(10)] // The higher the number the higher priority and we want this one last
-     internal class ConsoleDebugTargetsProvider : IDebugProfileLaunchTargetsProvider
+    internal class ConsoleDebugTargetsProvider : IDebugProfileLaunchTargetsProvider
     {
+        private static readonly ICollection<char> s_escapedChars = new List<char>()
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Debug/ConsoleDebugTargetsProvider.cs]: Nit, for memory/perf this was better as an array and typed as an array.

**Comment by MattGertz** [comment]
(Please retag me once changes are applied & CRs are done.)

**Comment by davkean** [comment]
👍

**Comment by 333fred** [comment]
@MattGertz ready for sign-off.

**Pull request merged**